src/librados/librados_asio.h

```diff
@@ -94,11 +97,48 @@ struct AsyncOp : Invoker<Result> {
     op.dispatch(std::move(p), ec, ver);
   }
 
+  struct op_cancellation {
+    IoCtx io;
```

Comment:
> we don't have a way to get at the `IoCtxImpl*` inside `IoCtx`

it is possible to get at it through `AioCompletion->pc->io`, but this doesn't help. we can't call functions on `IoCtxImpl` or `AioCompletionImpl` from this header because they aren't exposed by librados.so as public symbols:
> undefined reference to 'librados::IoCtxImpl::aio_cancel(librados::AioCompletionImpl*)'
> undefined reference to 'librados::AioCompletionImpl::cancel()'

Reply:
> the given `IoCtx` instance may get moved in the meantime, so we can't expect a pointer/reference to be stable

Is this something that ASIO "bindings" do?

Speaking generally, I think this is a grey area at best.  `IoCtx` has a move constructor, but whether it's allowed to be invoked with I/O in-flight isn't specified anywhere and the default answer would be no, at least in my book.  It seems similar to performing a destructive action on e.g. some library handle while it's still in use.

Reply:
i raised another pr in https://github.com/ceph/ceph/pull/60576 that also relates to lifetime of this IoCtx argument

Reply:
> asio has specific requirements for [Lifetime of initiating function arguments](https://www.boost.org/doc/libs/1_85_0/doc/html/boost_asio/reference/asynchronous_operations.html#boost_asio.reference.asynchronous_operations.lifetime_of_initiating_function_arguments) for async functions, and it bothers me that this `IoCtx& io` argument requires more of the caller than it should

Is it the ASIO requirement for lvalue reference to non-const parameters in general or something specifically about our `IoCtx` type that bothers you?  I'm double checking because the way I read the ASIO requirement, it matches my "no potentially destructive actions on the handle with I/O in flight" assumption and RGW usage seems to be in violation.

> 
> passing a stable `IoCtx&` often means heap-allocating one and managing its lifetime. but because IoCtx already does this for IoCtxImpl, i think users should be encoraged to rely on that instead

A stable `IoCtx` has never been a problem for me -- generally, there is one instance per pool (or RADOS namespace), and it's used throughout the program.  What necessitates the "copy `IoCtx` into memory that lives until completion and let the original one get destroyed" pattern in RGW?  Is it done on a per-I/O basis?  Does RGW recycle actual `IoCtxImpl` instances in some cases too?

Reply:
> Is it the ASIO requirement for lvalue reference to non-const parameters in general or something specifically about our `IoCtx` type that bothers you?

it's that `IoCtx` behaves like a `shared_ptr<IoCtxImpl>` without saying so. using the standardese language from that requirement, our implementation (including librados and Objecter) only assumes the validity of the pointee _inside_ the reference, not the reference itself. it's that requirement, not stability of `IoCtx&`, that i'd like to document in librados_asio.h instead

as you mentioned above, librados doesn't document the shallow-copy behavior of IoCtx well enough for callers to rely on it. what do you think about adding text like this?
```
   * IoCtx holds a pointer to its underlying implementation. The dup()
   * method performs a deep copy of this implementation, but the copy
   * construction and assignment operations perform shallow copies by
   * sharing that pointer.
   *
   * Function names starting with aio_ are asynchronous operations that
   * return immediately after submitting a request, and whose completions
   * are managed by the given AioCompletion pointer. The IoCtx's underlying
   * implementation is involved in the delivery of these completions, so
   * the caller must guarantee that its lifetime is preserved until then -
   * if not by preserving the IoCtx instance that submitted the request,
   * then by a copied/moved instance that shares the same implementation.
```

Reply:
I think that's a wonderful idea. It makes formal the behavior we have and aren't likely to change any time soon.